### PR TITLE
wipe all references lxc.network.{[i]}.* and replace with lxc.net.[i].*

### PR DESCRIPTION
**Pull request by 0x0916**
Fix: https://github.com/lxc/lxc/issues/1641

This pr also add the new `lxc.net` key support for function `network_new_hwaddrs`.

### CODE REVIEW

**Comment by 0x0916** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by brauner** [comment]
Note that the new network parse only supports indexed keys. That means you e.g. cannot replace `lxc.network.type` with `lxc.net.type`. You rather need to replace it with `lxc.net.0.type`.

**Comment by 0x0916** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by 0x0916** [comment]
@brauner Can you tell me that how to run testsuite on my laptop, or update the instructions to README?

**Comment by brauner** [comment]
The easiest way, assuming you have all build dependencies, is to grab the compile instructions from the jenkins output, and compiling locally. Then in a shell set `LD_LIBRARY_PATH` to the appropriate directory. This part (minus the compile flags) can be found in `README.md`. The you can simply run the tests from that shell (as root).

**Comment by 0x0916** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by 0x0916** [comment]
@brauner failed on i386 clang, do you know reason?

**Comment by brauner** [comment]
@0x0916 that looks like a false negative. Will restart jenkins to confirm.

**Comment by brauner** [comment]
Oh wait, it stopped before running the utility tests. Hm, let's see what the next jenkins run will do.

File: README.md [hunk 1]
```diff
@@ -83,8 +83,8 @@ LXC is configured via a simple set of keys. For example,
 - `lxc.mount.entry`
 
 LXC namespaces configuration keys by using single dots. This means complex
-configuration keys such as `lxc.network` expose various subkeys such as
-`lxc.network.type`, `lxc.network.link`, `lxc.network.ipv6`, and others for even
+configuration keys such as `lxc.net.0` expose various subkeys such as
+`lxc.net.0.type`, `lxc.net.0.link`, `lxc.net.0.ipv6`, and others for even
```
Comment (brauner) [inline README.md]: Excellent that you also thought of that. :)

File: src/tests/get_item.c [hunk 2]
```diff
@@ -77,12 +77,12 @@ int main(int argc, char *argv[])
 	}
```
Comment (brauner) [inline src/tests/get_item.c]: Actually, until `LXC 3.0` we should still run tests for the old configuration keys. So this file should only test for `lxc.network.{[i]}.*` related keys. The new network configuration keys are already tested in the `parse_config_file.c` test I added a while back.

File: src/tests/getkeys.c [hunk 3]
```diff
@@ -41,7 +41,7 @@ int main(int argc, char *argv[])
 		goto out;
```
Comment (brauner) [inline src/tests/getkeys.c]: Similar comment: until LXC 3.0 we should still run tests for the old configuration keys. So this file should only test for both `lxc.network.{[i]}.*` and `lxc.net.[i].*` keys.

File: src/lxc/confile_utils.c [hunk 4]
```diff
@@ -532,10 +532,44 @@ int rand_complete_hwaddr(char *hwaddr)
...
+		SYSERROR("failed to allocate memory");
+		return false;
+	}
+	if (*(copy + 8) >= '0' && *(copy + 8) <= '9') {
+		p = strchr(copy + 8, '.');
+		if (!p) {
+			free(copy);
+			return false;
+		}
+		strcpy(copy + 8, p + 1);
```
Comment (brauner) [inline src/lxc/confile_utils.c]: Hm, not a fan of `strcpy()` without bounds checking tbh.

File: src/lxc/confile_utils.c [hunk 5]
```diff
@@ -532,10 +532,44 @@ int rand_complete_hwaddr(char *hwaddr)
...
+	}
+	if (strncmp(copy, "lxc.net.hwaddr", 14) == 0) {
```
Comment (brauner) [inline src/lxc/confile_utils.c]: Isn't this supposed to handle bot `lxc.net.[i].hwaddr` and `lxc.network.{[i]}.hwaddr.*`? If so you should add:

```
if (strncmp(copy, "lxc.network.hwaddr", appropriate-length)  == 0) {
        /* same error handling */
}
```
You can probably move this into a single line by using `&&`.

File: src/lxc/confile_utils.c (hunk 4 above)
Comment (brauner) [inline src/lxc/confile_utils.c]: I'd prefer a

    strncpy(copy + 8, p + 1, length-you-want);
    copy[length-you-want - 1] = '\0';

or you use the same pattern but use `memmove()`. Sorry for the bikeshedding but I really don't approve of `strcpy()` That just calls for trouble later on.

Comment (0x0916) [inline src/lxc/confile_utils.c]: ok, just copy `hwaddr`(length = 6) is enough.

File: src/lxc/confile_utils.c (hunk 5 above)
Comment (0x0916) [inline src/lxc/confile_utils.c]: yes, in order to handle `lxc.network.{[i]}`, we need to dup another line and process it with the same logic.

File: src/tests/get_item.c (hunk 2 above)
Comment (0x0916) [inline src/tests/get_item.c]: updated

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
I think there's still an issue with retrieving:

    308: lxc.network.0.ipv4 returned -1

**Comment by brauner** [comment]
You could probably just copy and paste the old `get_item.c` file and see if it works correctly locally. I guess it doesn't need to be changed.

**Comment by 0x0916** [comment]
@brauner it not the problem of `get_item.c`. just revert the following change

```
diff --git a/config/etc/default.conf.lxcbr b/config/etc/default.conf.lxcbr
index e53047d..661718b 100644
--- a/config/etc/default.conf.lxcbr
+++ b/config/etc/default.conf.lxcbr
@@ -1,4 +1,4 @@
-lxc.network.type = veth
-lxc.network.link = lxcbr0
-lxc.network.flags = up
-lxc.network..hwaddr = 00:16:3e:xx:xx:xx
+lxc.net.0.type = veth
+lxc.net.0.link = lxcbr0
+lxc.net.0.flags = up
+lxc.net.0.hwaddr = 00:16:3e:xx:xx:xx
```

the testcase pass.

**Comment by 0x0916** [comment]
@brauner Maybe we can merge this pr, and solve `config/etc/default.conf.lxcbr` issue in another pr.

**Comment by brauner** [comment]
Oh ok, that makes perfect sense and isn't your fault. LXC will read in the `/etc/default.conf` if run as root and it won't find a defined network for `lxc.network.0.ipv4`. So what you can do is use `set_config_item()` and set

```
lxc.network.type = veth
lxc.network.link = lxcbr0
lxc.network.flags = up
lxc.network..hwaddr = 00:16:3e:xx:xx:xx
```

before performing the checks.

**Comment by 0x0916** [comment]
@brauner when I define a network before checks in `get_item.c`, There is still some error when clear the ipv4 config item. `c->clear_config_item(c, "lxc.network.0.ipv4")`.  I have fixed that. see https://github.com/lxc/lxc/pull/1649/commits/cf4b7999cd90c2f9c58dd1bc2217ad7cc254673a

**Comment by 0x0916** [comment]
jenkins: test this please

File: src/lxc/confile_legacy.c [hunk 6]
```diff
@@ -944,32 +944,31 @@ int get_config_network_legacy_item(const char *key, char *retv, int inlen,
...
-	lxc_list_for_each(it, &c->network) {
-		if (i == idx)
-			break;
-		i++;
 	}
-	if (i < idx)  // we don't have that many nics defined
-		return -1;
 
-	if (!it || !it->elem)
+	if ((idx = get_network_netdev_idx(key)) == EINVAL)
```
Comment (brauner) [inline src/lxc/confile_legacy.c]: Good catch.

**Comment by 0x0916** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Pull request merged**